src/follow-focus.cpp

```diff
@@ -0,0 +1,160 @@
+ * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+ * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE
+ * OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include <wayfire/plugin.hpp>
+#include <wayfire/core.hpp>
+#include <wayfire/geometry.hpp>
+#include <wayfire/option-wrapper.hpp>
+#include <wayfire/output.hpp>
+#include <wayfire/output-layout.hpp>
+#include <wayfire/workspace-manager.hpp>
+#include <wayfire/view.hpp>
+#include <wayfire/util.hpp>
+
+
+class wayfire_follow_focus : public wf::plugin_interface_t
+{
+  private:
+    wayfire_view last_view = nullptr;
+    wf::wl_timer change_focus;
+    wf::point_t last_coords;
+    double distance;
+
+    wf::option_wrapper_t<bool> should_change_view{"follow-focus/change_view"};
+    wf::option_wrapper_t<bool> should_change_output{"follow-focus/change_output"};
+    wf::option_wrapper_t<int> focus_delay{"follow-focus/focus_delay"};
+    wf::option_wrapper_t<int> threshold{"follow-focus/threshold"};
+    wf::option_wrapper_t<bool> raise_on_top{"follow-focus/raise_on_top"};
+
+    void change_view()
+    {
+        auto view = wf::get_core().get_cursor_focus_view();
+
+        if (raise_on_top)
+        {
+            wf::get_core().focus_view(view);
+        } else
+        {
+            wf::get_core().set_active_view(view);
+        }
+    }
+
+    void change_output()
+    {
+        auto coords = wf::get_core().get_cursor_position();
+        for (auto output : wf::get_core().output_layout->get_outputs())
+        {
+            if (output->get_layout_geometry() & coords)
+            {
+                wf::get_core().focus_output(output);
+
+                return;
+            }
+        }
+    }
+
+    void do_actions()
+    {
+        if (should_change_view)
+        {
+            change_view();
+        }
+
+        if (should_change_output)
+        {
+            change_output();
+        }
+
+        distance = -1;
+    }
+
+    wf::signal_callback_t pointer_motion = [=] (wf::signal_data_t* /*data*/)
+    {
+        change_focus.disconnect();
+
+        auto view = wf::get_core().get_cursor_focus_view();
+
+        if (!view || (view->role != wf::VIEW_ROLE_TOPLEVEL) ||
+            (output->workspace->get_view_layer(view) != wf::LAYER_WORKSPACE))
+        {
+            return;
+        }
+
+        if (view != last_view)
+        {
+            distance  = -1;
+            last_view = view;
+        }
+
+        /* Update how much the cursor moved this time */
+        auto cpf = wf::get_core().get_cursor_position();
+        wf::point_t coords{static_cast<int>(cpf.x), static_cast<int>(cpf.y)};
+        if (distance == -1)
+        {
+            distance = 0;
+        } else
+        {
+            distance += abs(coords - last_coords);
+        }
+
+        last_coords = coords;
```

Comment:
this really calculates the 'total mouse mouvement' i think it may be wrong,
what we want to implement is a 'fringe' i.e. if you enter the window more than x pixels you switch.
with this logic if you step into the window (say 10px) and then your mouse wobbles 3 pixels, you'll eventually have distance > 10 and switch focus.

I feel we should only update last_coords when first entering the window (see the state machine in #43)

Reply:
the user in #17 talked about his touchpad getting accidently hit and moving the cursor a tad, with the current implementation, a bit more wobble and the focus will be lost.

Reply:
> this really calculates the 'total mouse mouvement' i think it may be wrong,
> what we want to implement is a 'fringe' i.e. if you enter the window more than x pixels you switch.
> with this logic if you step into the window (say 10px) and then your mouse wobbles 3 pixels, you'll eventually have distance > 10 and switch focus.
> 
> I feel we should only update last_coords when first entering the window (see the state machine in #43)

I think you're right about this. I've already merged the plugin, did you want to submit a patch?

Reply:
sure.